### DISCUSSION TITLE
How to handle RawContacts that have multiple Notes in Oppo ColorOS?

### DISCUSSION THREAD

**Discussion by Yameteh** [vestrel00/contacts-android#342] [Q&A] [answered]
Hello,

I am working on integrating with the contacts-android, specifically concerning contact management features. I've encountered a scenario where a contact may have multiple notes assigned to them, although this is not a typical occurrence. However, upon fetching a contact's details via the current API implementation, it appears that only a single note is returned.

I would appreciate it if you could consider extending the API functionality to handle the less common but possible case where a contact has multiple notes associated with it. This enhancement could improve the robustness and adaptability of the contact management system.

Thanks for your attention to this matter, and I look forward to your response.

Best regards

**Comment by vestrel00** [comment]
Hey! Thanks for using these APIs! 

So, I wrote the entities such that a **RawContact** can only have one note. This is the same behavior that you see across all other contacts apps such as AOSP contacts and Google contacts. Even the iOS contacts app only allows one note per contact.

> I've encountered a scenario where a contact may have multiple notes assigned to them

Keep in mind that a **RawContact** is different from a **Contact**. A **Contact** may have one or more **RawContact**s associated with it. In this case, the **Contact** may have more than one Note if it has two or more **RawContact**s each having a non-blank Note. **Contact**s having multiple Notes is already supported by the current APIs. Perhaps, this is the scenario you encountered?

Please read this page of the docs; https://github.com/vestrel00/contacts-android/blob/main/docs/entities/about-api-entities.md

![image](https://github.com/vestrel00/contacts-android/assets/4332753/1bce2782-62a9-4259-87d9-833aa11c7d06)

In case you are new to contacts management, I strongly suggest reading the entire page I linked above so you can understand the one to many relationship between Contact -> RawContact -> Data (e.g. Note)

----------------

If you believe the API should support more than one Note per **RawContact**, could you please attach a video of the occurrence? Can you reproduce it in the AOSP or Google Contacts app?

When you open the AOSP or Google Contacts app and click on a **Contact**, it will display the aggregate **Contact** in the details screen. If that **Contact** has two or more **RawContact**s linked to it, then the details screen may show two or more Notes. If you try to edit that **Contact**, you will see that each **RawContact** can only have one Note.

Did this answer your question? Let me know, thanks! 🙏

**Comment by Yameteh** [comment]
> Hey! Thanks for using these APIs!
> 
> So, I wrote the entities such that a **RawContact** can only have one note. This is the same behavior that you see across all other contacts apps such as AOSP contacts and Google contacts. Even the iOS contacts app only allows one note per contact.
> 
> > I've encountered a scenario where a contact may have multiple notes assigned to them
> 
> Keep in mind that a **RawContact** is different from a **Contact**. A **Contact** may have one or more **RawContact**s associated with it. In this case, the **Contact** may have more than one Note if it has two or more **RawContact**s each having a non-blank Note. **Contact**s having multiple Notes is already supported by the current APIs. Perhaps, this is the scenario you encountered?
> 
> Please read this page of the docs; https://github.com/vestrel00/contacts-android/blob/main/docs/entities/about-api-entities.md
> 
> ![image](https://private-user-images.githubusercontent.com/4332753/316018961-1bce2782-62a9-4259-87d9-833aa11c7d06.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzMzU0OTQsIm5iZiI6MTcxMTMzNTE5NCwicGF0aCI6Ii80MzMyNzUzLzMxNjAxODk2MS0xYmNlMjc4Mi02MmE5LTQyNTktODdkOS04MzNhYTExYzdkMDYuanBlZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI1VDAyNTMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjMjRmYTAzZjVlMjBmYzJhOGU5YTkwZDljMzczMzc0MTgxY2RkMjk3NDIyYzUxNTFlOGI1MDI3MDJjNTlkZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UFXK4Jmn8l-xDur0PgMQmqWHnyZjnWu9UFjtEjQHuIk)
> 
> In case you are new to contacts management, I strongly suggest reading the entire page I linked above so you can understand the one to many relationship between Contact -> RawContact -> Data (e.g. Note)
> 
> If you believe the API should support more than one Note per **RawContact**, could you please attach a video of the occurrence? Can you reproduce it in the AOSP or Google Contacts app?
> 
> When you open the AOSP or Google Contacts app and click on a **Contact**, it will display the aggregate **Contact** in the details screen. If that **Contact** has two or more **RawContact**s linked to it, then the details screen may show two or more Notes. If you try to edit that **Contact**, you will see that each **RawContact** can only have one Note.
> 
> Did this answer your question? Let me know, thanks! 🙏

So far I've only found the ability to enter multiple notes inside a single rawContact on oppo coloros, it doesn't appear on the AOSP or Google Contacts apps. Would you be open to considering this API modification?
Best regards
![Screenshot_20240325_105551](https://github.com/vestrel00/contacts-android/assets/13945256/25b7e1d9-6bcd-4417-a550-3ca88d3ac6e3)

**Comment by vestrel00** [comment]
> oppo coloros

Hmmm... I do not have an Oppo phone. I wonder if those notes are [custom data mime types](https://github.com/vestrel00/contacts-android/blob/main/docs/customdata/integrate-custom-data-from-other-apps.md). Oppo might not be using the standard note mime type that AOSP and Google Contacts apps are using. OR, Oppo is not following the rule that there should only be one note per RawContact.

Let's see if we can figure this out! I need you to do the following things for me.

1. Create a RawContact in Oppo with 2 or more notes. 
2. Then, in your own app, use this library to log the Contacts Provider table so we can see the mimetype that Oppo is using for those notes. https://github.com/vestrel00/contacts-android/blob/main/docs/debug/debug-contacts-provider-tables.md
3. Please use test data and paste it as a reply in this issue. Do not paste real data! It will be public!
4. Additionally, view that same RawContact in the Google Contacts or AOSP Contacts app and screenshot what you see. Do you see the multiple notes? Do you only see one note? Do you see zero notes?

Please let me know!

**Comment by vestrel00** [comment]
> Would you be open to considering this API modification?

It depends on how Oppo implemented notes. Let's figure that out first before I answer this question 😃

**Comment by Yameteh** [comment]
> > oppo coloros
> 
> Hmmm... I do not have an Oppo phone. I wonder if those notes are [custom data mime types](https://github.com/vestrel00/contacts-android/blob/main/docs/customdata/integrate-custom-data-from-other-apps.md). Oppo might not be using the standard note mime type that AOSP and Google Contacts apps are using. OR, Oppo is not following the rule that there should only be one note per RawContact.
> 
> Let's see if we can figure this out! I need you to do the following things for me.
> 
> 1. Create a RawContact in Oppo with 2 or more notes.
> 2. Then, in your own app, use this library to log the Contacts Provider table so we can see the mimetype that Oppo is using for those notes. https://github.com/vestrel00/contacts-android/blob/main/docs/debug/debug-contacts-provider-tables.md
> 3. Please use test data and paste it as a reply in this issue. Do not paste real data! It will be public!
> 4. Additionally, view that same RawContact in the Google Contacts or AOSP Contacts app and screenshot what you see. Do you see the multiple notes? Do you only see one note? Do you see zero notes?
> 
> Please let me know!

contacts provider table log 
[OPPO_CONTACTS.txt](https://github.com/vestrel00/contacts-android/files/14753953/OPPO_CONTACTS.txt)

**Comment by vestrel00** [comment]
Thanks for the logs! I don't see any custom data there. Can you also do one more thing.

View/edit RawContact that has two notes in the Google Contacts or AOSP Contacts app and screenshot what you see. Do you see the multiple notes? Do you only see one note? Do you see zero notes?

**Comment by Yameteh** [comment]
@vestrel00 
google contacts app ui display the multiple notes

![Screenshot_20240326_155449](https://github.com/vestrel00/contacts-android/assets/13945256/6ae26feb-6301-499f-9c89-b3c09fc038f9)

**Comment by vestrel00** [comment]
I do not see any of the contact data in your screenshot that is also in the log file.

You could please attach a log file and a screenshot that corresponds to the log file?

Thanks for your cooperation 🙏

**Comment by Yameteh** [comment]
> I do not see any of the contact data in your screenshot that is also in the log file.
> 
> You could please attach a log file and a screenshot that corresponds to the log file?
> 
> Thanks for your cooperation 🙏

[OPPO_CONTACTS(1).txt](https://github.com/vestrel00/contacts-android/files/14755174/OPPO_CONTACTS.1.txt)

**Comment by vestrel00** [comment]
It seems like all of those notes do belong to the same RawContact and that Google Contacts app is displaying all of the notes in view-mode.

Here are my last questions before I give you an answer. In **Google Contacts** app...

1. What happens when you **edit** a RawContact with more than one note (what do you see when you are in edit-mode)?
2. Does Google Contacts app allow you to edit all of the notes or only one?
3. Are you allowed to add another note or remove an existing note? 
4. What happens when you save your edits? Do you still see multiple notes in view-mode?

Those should be my final questions. Once you answer them, I will look into what to do about supporting multiple notes using this library. Thanks for your cooperation! 

_Note to self, could this be a new feature in the latest version of Android? Or is it really just Oppo ColorOS?_

**Comment by Yameteh** [comment]
> It seems like all of those notes do belong to the same RawContact and that Google Contacts app is displaying all of the notes in view-mode.
> 
> Here are my last questions before I give you an answer. In **Google Contacts** app...
> 
> 1. What happens when you **edit** a RawContact with more than one note (what do you see when you are in edit-mode)?
> 2. Does Google Contacts app allow you to edit all of the notes or only one?
> 3. Are you allowed to add another note or remove an existing note?
> 4. What happens when you save your edits? Do you still see multiple notes in view-mode?
> 
> Those should be my final questions. Once you answer them, I will look into what to do about supporting multiple notes using this library. Thanks for your cooperation!
> 
> _Note to self, could this be a new feature in the latest version of Android? Or is it really just Oppo ColorOS?_

2. I could edit all of the notes
3. I could remove an existing note but not add one 
4. I could still see multiple notes after the edits saved

**Comment by vestrel00** [comment]
Thanks! That's all the info I need. I think I should be able to write extension functions to support your use case of adding/editing/removing multiple Notes per RawContact. 

I will keep the default behavior of one Note per RawContact. The multiple notes behavior will be opt-in.

I'm a bit busy at work so this might take a while to do. I'll try to get it done some time in April. This will be part of the upcoming release; [0.3.2](https://github.com/vestrel00/contacts-android/discussions/316) 😁

**Comment by vestrel00** [comment]
I will convert this to a discussion and then create an issue for the new API.

**Comment by vestrel00** [comment, marked as answer]
@Yameteh, I created issue #343 ! It will be part of the upcoming release; [0.3.2](https://github.com/vestrel00/contacts-android/discussions/316). I'll try to get it out some time in April.

**Reply by vestrel00** [reply]
Change of plans due to new issues piling up and work and life still being too busy.

This (#343) will not be included in the upcoming release [0.3.2](https://github.com/vestrel00/contacts-android/discussions/316) but it will be included in the follow-up release [0.3.3](https://github.com/vestrel00/contacts-android/discussions/349)

Apologies for the delays 🙏
